Use new sys.thread package when possible on haxe 4

Aidan63 opened the pull request:
Hello,
Haxe 4.0.0-rc.2 introduces the `sys.thread` package to replace the individual threading classes for interp, neko, cpp, hl, and java and I've gone through and made it use those classes when possible.
Hopefully I've found all the important conditional compilation sections, it's working for my use on haxe 4 but I don't use all the targets, hopefully the ci will catch any issues!

Aidan63 committed:
Use new haxe 4 experimental threads where possible

simplified threads conditional compilation

Updated RLock to use sys.thread.Mutex when on haxe 4

Updated queue to use sys.thread.Deque when on haxe 4

Updated #if syntax to make older haxe versions happy

Simplified java and haxe version conditionals

sebthom commented:
Hi Aidan, 
thanks for your PR. Unfortunately your changes seem to break the library on Haxe 3.4.7 as you can see in the build logs at https://travis-ci.org/vegardit/haxe-concurrent/jobs/510055482

For the PR to be accepted please fix the syntax error and squash your commits.

Thank you!

Seb

Aidan63 commented:
Hi (and sorry for the late reply), was meaning to sort out the issues over the weekend but forgot!

I've fixed up what appears to be the last few remaining issues and squashed everything down to a single commit, so once this CI round is done hopefully it should be good to go.

sebthom commented:
Thanks a lot!
